### PR TITLE
[docs] #39, #83, #90: Update WASM/FFI guide

### PR DESCRIPTION
**Pull request by outoftardis**

### CODE REVIEW

File: src/guide/advanced/ffi.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Foreign Function Interfaces (FFI)
+
+As we mentioned before, Iroha 2 relies on the C-linkage to generate
+[WASM](./wasm.md) bindings, and the `iroha_wasm` crate deals with foreign
+function interfaces (FFI). Here we discuss in more details how Iroha 2
```
Comment (WRRicht3r) [inline src/guide/advanced/ffi.md:5]:
```suggestion
function interfaces (FFI). Here we will discuss in more detail how Iroha 2
```

File: src/guide/advanced/ffi.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+handles FFI logic.
+
+## FFI Binding Generation
+
+The `iroha_ffi` crate is used to generate functions that are callable via
+FFI.
+
+A Rust structure is converted into an intermediary FFI struct via `IntoFfi`
+trait, which implements the `AsReprCRef` trait and can be converted into a
```
Comment (WRRicht3r) [inline src/guide/advanced/ffi.md:14]:
```suggestion
trait, which implements the `AsReprCRef` trait, and can be converted into a
```

File: src/guide/advanced/ffi.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+[separate section](#name-mangling)).
+
+![Untitled](/img/ffi.png)
+
+The traits that enable binding generation are `ReprC`, `AsReprCRef`,
+`TryFromReprC`, and `IntoFfi`:
+
+| Trait          | Description                                                                                                       |
+| -------------- | ----------------------------------------------------------------------------------------------------------------- |
+| `ReprC`        | The trait represents the robust type that conforms to C ABI. The type can be safely shared across FFI boundaries. |
```
Comment (WRRicht3r) [inline src/guide/advanced/ffi.md:29]:
```suggestion
| `ReprC`        | This trait represents the robust type that conforms to C ABI. The type can be safely shared across FFI boundaries. |
```

File: src/guide/advanced/ffi.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+| `AsReprCRef`   | The trait is used to convert from a Rust type reference to `ReprC` reference.                                     |
```
Comment (WRRicht3r) [inline src/guide/advanced/ffi.md:30]:
```suggestion
| `AsReprCRef`   | This trait is used to convert from a Rust type reference to `ReprC` reference.                                     |
```

File: src/guide/advanced/ffi.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+| `TryFromReprC` | The trait is used to perform conversion from a type that implements `ReprC`.                                      |
```
Comment (WRRicht3r) [inline src/guide/advanced/ffi.md:31]:
```suggestion
| `TryFromReprC` | This trait is used to perform a conversion from a type that implements `ReprC`.                                      |
```

File: src/guide/advanced/ffi.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+| `IntoFfi`      | The trait is used to convert into a type that can be converted to an FFI-compatible `ReprC` type.                 |
```
Comment (WRRicht3r) [inline src/guide/advanced/ffi.md:32]:
```suggestion
| `IntoFfi`      | This trait is used to convert into a type that can then be converted to an FFI-compatible `ReprC` type.                 |
```

**Review by WRRicht3r — DISMISSED** [review]

File: src/guide/advanced/ffi.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+- For the `trait_method_name` method in `StructName` struct, the FFI name
+  would be `StructName__trait_method_name`.
+- For setting the `field_name` field in `StructName` struct, the FFI name
+  would be `StrucuName__set_field_name`.
+- For getting the `field_name` field in `StructName` struct, the FFI name
+  would be `StrucuName__field_name`.
+- For getting the mutable `field_name` field in `StructName` struct, the
+  FFI name would be `StrucuName__field_name_mut`.
+- For the `method_name` method in `module_name`, the FFI name would be
+  `module__name::__method_name`.
```
Comment (mversic) [inline src/guide/advanced/ffi.md:52]:
```suggestion
- For the freestanding `fn_name` function in `module_name`, the FFI name would be
  `module_name::__fn_name`.
```

there are two things, freestanding functions as above and then there are some traits that are not generic and that allow for sharing of implementation in the FFI
```rs
impl Clone for Type1 {
    fn clone(&self) -> Self;
}
impl Clone for Type2 {
    fn clone(&self) -> Self;
}
```

for which `module_name::__clone` will be generated in the same regard as for freestanding functions

**Review by mversic — DISMISSED** [review]

File: src/guide/advanced/ffi.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+types.
+
+### Name Mangling
+
+Note the use of double underscores in generated names of FFI objects:
+
+- For the `inherent_fn` function in the `StructName` struct, the FFI name
+  would be `StructName__inherent_fn`.
+- For the `trait_method_name` method in `StructName` struct, the FFI name
+  would be `StructName__trait_method_name`.
```
[the remark this answers is not in the document]
Comment (Erigara) [reply src/guide/advanced/ffi.md:44]: Yes

File: src/guide/advanced/ffi.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+[WASM](./wasm.md) bindings, and the `iroha_wasm` crate deals with foreign
+function interfaces (FFI). Here we will discuss in more detail how Iroha 2
+handles FFI logic.
+
+## FFI Binding Generation
+
+The `iroha_ffi` crate is used to generate functions that are callable via
+FFI.
+
+A Rust structure is converted into an intermediary FFI struct via `IntoFfi`
```
Comment (WRRicht3r) [inline src/guide/advanced/ffi.md:13]:
```suggestion
A Rust structure is converted into an intermediary FFI struct via the `IntoFfi`
```

**Review by WRRicht3r — CHANGES_REQUESTED** [review]

File: src/guide/diagrams-src/ffi.tex [hunk 10]
```diff
@@ -0,0 +1,73 @@
+\documentclass[tikz,border={0 1}]{standalone}
```
Comment (0x009922) [inline src/guide/diagrams-src/ffi.tex:1]: There is a [`markdown-it-texmath`](https://www.npmjs.com/package/markdown-it-texmath) package that could help with TeX rendering directly from source, but... I tried it's [demo](https://goessner.github.io/markdown-it-texmath/index.html) and didn't find a way how to display this file.

**Review by 0x009922 — APPROVED** [review]

File: src/guide/diagrams-src/ffi.tex (hunk 10 above)
Comment (outoftardis) [reply src/guide/diagrams-src/ffi.tex:1]: thanks for the info! if we have more diagrams, we might want to look into it to automate the rendering

**Review by appetrosyan — APPROVED** [review]

File: src/guide/advanced/ffi.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
+# Foreign Function Interfaces (FFI)
+
+As we mentioned before, Iroha 2 relies on the C-linkage to generate
+[WASM](./wasm.md) bindings, and the `iroha_wasm` crate deals with foreign
+function interfaces (FFI). Here we will discuss in more detail how Iroha 2
+handles FFI logic.
```
[the remark this answers is not in the document]
Comment (outoftardis) [reply src/guide/advanced/ffi.md:6]: applied in #158

File: src/guide/advanced/ffi.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+## FFI Binding Generation
+
+The `iroha_ffi` crate is used to generate functions that are callable via
+FFI.
```
[the remark this answers is not in the document]
Comment (outoftardis) [reply src/guide/advanced/ffi.md:11]: applied in #158

**Pull request merged**